Shipper Doesn't Work With Resyncs Disabled

If you run Shipper with resyncs disabled, there are cases where Shipper will get stuck while taking a release forward.

This is because sometimes we rely on the resync to happen, and we don't do as much as we can in each sync loop.

We haven't dug into this yet, so the best way to reproduce this is to run the end-to-end tests while running Shipper with the resync period of 0 to disable all resyncs, and go from there.

I followed the suggestion of just running the end-to-end tests. Much to my surprise, they all passed. Unfortunately, that seems to be because in current master there's no way to completely disable resyncs. I fixed that in 7ae1c636b93eeb680faea8fedfa69e228dd43ad6.

With those changes, I get one failing test:

```
--- FAIL: TestRolloutAbort (73.78s)
    e2e_test.go:751: waiting for release "my-test-app-77cf3473-0" took 1.597339649s
    e2e_test.go:840: waiting for completion of "my-test-app-77cf3473-0" took 18.399953573s
    e2e_test.go:617: waiting for contender release to appear after editing app "my-test-app"
    e2e_test.go:751: waiting for release "my-test-app-3037a9fd-0" took 1.598389769s
    e2e_test.go:624: setting contender release "my-test-app-3037a9fd-0" targetStep to 0
    e2e_test.go:627: waiting for contender release "my-test-app-3037a9fd-0" to achieve waitingForCommand for targetStep 0
    e2e_test.go:761: release strategy state transition: "" -> "release \"my-test-app-3037a9fd-0\" has no achievedStep reported yet"
    e2e_test.go:761: release strategy state transition: "release \"my-test-app-3037a9fd-0\" has no achievedStep reported yet" -> "{installation: False, capacity: False, traffic: False, command: True}"
    e2e_test.go:819: waiting for command took 10.398270118s
    e2e_test.go:633: checking that incumbent "my-test-app-77cf3473-0" has 4 pods and contender "my-test-app-3037a9fd-0" has 1 pods (strategy step 0 -- 100/1)
    e2e_test.go:624: setting contender release "my-test-app-3037a9fd-0" targetStep to 1
    e2e_test.go:627: waiting for contender release "my-test-app-3037a9fd-0" to achieve waitingForCommand for targetStep 1
    e2e_test.go:761: release strategy state transition: "" -> "{installation: False, capacity: True, traffic: False, command: False}"
    e2e_test.go:761: release strategy state transition: "{installation: False, capacity: True, traffic: False, command: False}" -> "{installation: False, capacity: False, traffic: True, command: False}"
    e2e_test.go:761: release strategy state transition: "{installation: False, capacity: False, traffic: True, command: False}" -> "{installation: False, capacity: False, traffic: False, command: True}"
    e2e_test.go:819: waiting for command took 6.397543921s
    e2e_test.go:633: checking that incumbent "my-test-app-77cf3473-0" has 2 pods and contender "my-test-app-3037a9fd-0" has 2 pods (strategy step 1 -- 50/50)
    e2e_test.go:761: release strategy state transition: "" -> "{installation: False, capacity: False, traffic: False, command: False}"
    e2e_test.go:814: timed out waiting for release to be waiting for capacity: waited 30s. final state: {installation: False, capacity: False, traffic: False, command: False}
FAIL
FAIL    github.com/bookingcom/shipper/test/e2e  104.361s
```

This particular instance happens because the strategy controller doesn't watch for releases being deleted, so the rollout abort gets stuck thinking it still has capacity. A cursory look also shows that most controllers don't handle deleted resources either, so even if this particular broken test gets fixed it doesn't mean we have fixed the entire problem.

--- a/cmd/shipper/main.go
+++ b/cmd/shipper/main.go
@@ -55,6 +55,7 @@ var controllers = []string{
 }
 
 const defaultRESTTimeout time.Duration = 10 * time.Second
+const defaultResync time.Duration = 30 * time.Second
 
 var (
 	masterURL           = flag.String("master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
@@ -62,12 +63,12 @@ var (
 	certPath            = flag.String("cert", "", "Path to the TLS certificate for target clusters.")
 	keyPath             = flag.String("key", "", "Path to the TLS private key for target clusters.")
 	ns                  = flag.String("namespace", shipper.ShipperNamespace, "Namespace for Shipper resources.")
-	resyncPeriod        = flag.String("resync", "30s", "Informer's cache re-sync in Go's duration format.")
 	enabledControllers  = flag.String("enable", strings.Join(controllers, ","), "comma-seperated list of controllers to run (if not all)")
 	disabledControllers = flag.String("disable", "", "comma-seperated list of controllers to disable")
 	workers             = flag.Int("workers", 2, "Number of workers to start for each controller.")
 	metricsAddr         = flag.String("metrics-addr", ":8889", "Addr to expose /metrics on.")
 	chartCacheDir       = flag.String("cachedir", filepath.Join(os.TempDir(), "chart-cache"), "location for the local cache of downloaded charts")
+	resync              = flag.Duration("resync", defaultResync, "Informer's cache re-sync in Go's duration format.")
 	restTimeout         = flag.Duration("rest-timeout", defaultRESTTimeout, "Timeout value for management and target REST clients. Does not affect informer watches.")
 )
 
@@ -87,7 +88,7 @@ type cfg struct {
 
 	kubeInformerFactory    informers.SharedInformerFactory
 	shipperInformerFactory shipperinformers.SharedInformerFactory
-	resync                 time.Duration
+	resync                 *time.Duration
 
 	recorder func(string) record.EventRecorder
 
@@ -107,11 +108,6 @@ type cfg struct {
 func main() {
 	flag.Parse()
 
-	resync, err := time.ParseDuration(*resyncPeriod)
-	if err != nil {
-		glog.Fatal(err)
-	}
-
 	baseRestCfg, err := clientcmd.BuildConfigFromFlags(*masterURL, *kubeconfig)
 	if err != nil {
 		glog.Fatal(err)
@@ -126,8 +122,8 @@ func main() {
 	stopCh := setupSignalHandler()
 	metricsReadyCh := make(chan struct{})
 
-	kubeInformerFactory := informers.NewSharedInformerFactory(informerKubeClient, resync)
-	shipperInformerFactory := shipperinformers.NewSharedInformerFactory(informerShipperClient, resync)
+	kubeInformerFactory := informers.NewSharedInformerFactory(informerKubeClient, *resync)
+	shipperInformerFactory := shipperinformers.NewSharedInformerFactory(informerShipperClient, *resync)
 
 	shipperscheme.AddToScheme(scheme.Scheme)
 
@@ -173,6 +169,7 @@ func main() {
 		shipperInformerFactory.Shipper().V1alpha1().Clusters(),
 		*ns,
 		restTimeout,
+		resync,
 	)
 
 	wg.Add(1)

--- a/pkg/clusterclientstore/store.go
+++ b/pkg/clusterclientstore/store.go
@@ -30,6 +30,7 @@ type Store struct {
 	ns          string
 	buildClient ClientBuilderFunc
 	restTimeout *time.Duration
+	resync      *time.Duration
 	cache       cache.CacheServer
 
 	secretInformer  corev1informer.SecretInformer
@@ -55,12 +56,14 @@ func NewStore(
 	secretInformer corev1informer.SecretInformer,
 	clusterInformer shipperinformer.ClusterInformer,
 	ns string,
-	restTimeout *time.Duration,
+	restTimeout,
+	resync *time.Duration,
 ) *Store {
 	s := &Store{
 		ns:          ns,
 		buildClient: buildClient,
 		restTimeout: restTimeout,
+		resync:      resync,
 		cache:       cache.NewServer(),
 
 		secretInformer:  secretInformer,
@@ -277,8 +280,12 @@ func (s *Store) create(cluster *shipper.Cluster, secret *corev1.Secret) error {
 		return fmt.Errorf("create informer client for Cluster %q: %s", cluster.Name, err)
 	}
 
-	// TODO(asurikov): propagate -resync from cmd/shipper here.
-	informerFactory := kubeinformers.NewSharedInformerFactory(informerClient, time.Second*30)
+	var resyncPeriod time.Duration
+	if s.resync != nil {
+		resyncPeriod = *s.resync
+	}
+
+	informerFactory := kubeinformers.NewSharedInformerFactory(informerClient, resyncPeriod)
 	// Register all the resources that the controllers are interested in, e.g.
 	// informerFactory.Core().V1().Pods().Informer().
 	for _, cb := range s.subscriptionRegisterFuncs {

--- a/pkg/clusterclientstore/store_test.go
+++ b/pkg/clusterclientstore/store_test.go
@@ -263,7 +263,7 @@ func (f *fixture) newStore() (*Store, kubeinformers.SharedInformerFactory, shipp
 	f.kubeClient = kubefake.NewSimpleClientset(f.kubeObjects...)
 	f.shipperClient = shipperfake.NewSimpleClientset(f.shipperObjects...)
 
-	const noResyncPeriod time.Duration = 0
+	noResyncPeriod := time.Duration(0)
 	shipperInformerFactory := shipperinformers.NewSharedInformerFactory(f.shipperClient, noResyncPeriod)
 	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(f.kubeClient, noResyncPeriod)
 
@@ -275,6 +275,7 @@ func (f *fixture) newStore() (*Store, kubeinformers.SharedInformerFactory, shipp
 		shipperInformerFactory.Shipper().V1alpha1().Clusters(),
 		shipper.ShipperNamespace,
 		f.restTimeout,
+		&noResyncPeriod,
 	)
 
 	return store, kubeInformerFactory, shipperInformerFactory

--- a/pkg/controller/capacity/capacity_controller.go
+++ b/pkg/controller/capacity/capacity_controller.go
@@ -182,6 +182,7 @@ func (c *Controller) capacityTargetSyncHandler(key string) bool {
 
 	ct = ct.DeepCopy()
 
+	shouldRetry := false
 	targetNamespace := ct.Namespace
 	selector := labels.Set(ct.Labels).AsSelector()
 
@@ -216,6 +217,7 @@ func (c *Controller) capacityTargetSyncHandler(key string) bool {
 		targetDeployment, err := c.findTargetDeploymentForClusterSpec(clusterSpec, targetNamespace, selector, clusterStatus)
 		if err != nil {
 			c.recordErrorEvent(ct, err)
+			shouldRetry = true
 			continue
 		}
 
@@ -228,6 +230,7 @@ func (c *Controller) capacityTargetSyncHandler(key string) bool {
 			_, err = c.patchDeploymentWithReplicaCount(targetDeployment, clusterSpec.Name, replicaCount, clusterStatus)
 			if err != nil {
 				c.recordErrorEvent(ct, err)
+				shouldRetry = true
 				continue
 			}
 		}
@@ -274,7 +277,7 @@ func (c *Controller) capacityTargetSyncHandler(key string) bool {
 		return true
 	}
 
-	return false
+	return shouldRetry
 }
 
 func (c *Controller) enqueueCapacityTarget(obj interface{}) {

--- a/pkg/controller/installation/installation_controller.go
+++ b/pkg/controller/installation/installation_controller.go
@@ -236,18 +236,6 @@ func (c *Controller) processInstallation(it *shipper.InstallationTarget) error {
 		return nil
 	}
 
-	if appLabelValue, ok := release.GetLabels()[shipper.AppLabel]; !ok {
-		// TODO(isutton): Transform this into a real error
-		return fmt.Errorf("couldn't find label %q in release %q", shipper.AppLabel, release.GetName())
-	} else if app, appListerErr := c.appLister.Applications(release.GetNamespace()).Get(appLabelValue); appListerErr != nil {
-		// TODO(isutton): wrap this error.
-		return appListerErr
-	} else if len(app.Status.History) > 0 && app.Status.History[len(app.Status.History)-1] != release.GetName() {
-		// Current release isn't the contender, so we do not attempt to
-		// create or modify objects at all.
-		return nil
-	}
-
 	installer := NewInstaller(c.chartFetchFunc, release, it)
 
 	// Build .status over based on the current .spec.clusters.

--- a/pkg/controller/installation/installation_controller.go
+++ b/pkg/controller/installation/installation_controller.go
@@ -149,7 +149,7 @@ func (c *Controller) processNextWorkItem() bool {
 		if c.workqueue.NumRequeues(key) >= maxRetries {
 			// Drop the InstallationTarget's key out of the workqueue and thus reset its
 			// backoff. This limits the time a "broken" object can hog a worker.
-			glog.Warningf("CapacityTarget %q has been retried too many times, dropping from the queue", key)
+			glog.Warningf("InstallationTarget %q has been retried too many times, dropping from the queue", key)
 			c.workqueue.Forget(key)
 
 			return true

--- a/pkg/controller/release/release_controller.go
+++ b/pkg/controller/release/release_controller.go
@@ -138,6 +138,7 @@ func NewController(
 			UpdateFunc: func(oldObj, newObj interface{}) {
 				controller.enqueueRelease(newObj)
 			},
+			DeleteFunc: controller.enqueueAppFromRelease,
 		})
 
 	installationTargetInformer.Informer().AddEventHandler(
@@ -146,6 +147,7 @@ func NewController(
 			UpdateFunc: func(oldObj, newObj interface{}) {
 				controller.enqueueInstallationTarget(newObj)
 			},
+			DeleteFunc: controller.enqueueInstallationTarget,
 		})
 
 	capacityTargetInformer.Informer().AddEventHandler(
@@ -154,6 +156,7 @@ func NewController(
 			UpdateFunc: func(oldObj, newObj interface{}) {
 				controller.enqueueCapacityTarget(newObj)
 			},
+			DeleteFunc: controller.enqueueCapacityTarget,
 		})
 
 	trafficTargetInformer.Informer().AddEventHandler(
@@ -162,6 +165,7 @@ func NewController(
 			UpdateFunc: func(oldObj, newObj interface{}) {
 				controller.enqueueTrafficTarget(newObj)
 			},
+			DeleteFunc: controller.enqueueTrafficTarget,
 		})
 
 	return controller
@@ -442,6 +446,22 @@ func (c *Controller) enqueueRelease(obj interface{}) {
 	}
 
 	c.releaseWorkqueue.Add(key)
+}
+
+func (c *Controller) enqueueAppFromRelease(obj interface{}) {
+	rel, ok := obj.(*shipper.Release)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("not a shipper.Release: %#v", obj))
+		return
+	}
+
+	appName, err := c.getAssociatedApplicationKey(rel)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("error fetching Application key for release %v: %s", rel, err))
+		return
+	}
+
+	c.applicationWorkqueue.Add(appName)
 }
 
 func (c *Controller) enqueueInstallationTarget(obj interface{}) {

--- a/pkg/testing/clusterclientstore.go
+++ b/pkg/testing/clusterclientstore.go
@@ -49,6 +49,7 @@ func ClusterClientStore(
 		shipperInformerFactory.Shipper().V1alpha1().Clusters(),
 		TestNamespace,
 		nil,
+		nil,
 	)
 
 	kubeInformerFactory.Start(stopCh)

--- a/test/e2e/e2e_test.go
+++ b/test/e2e/e2e_test.go
@@ -621,10 +621,10 @@ func TestRolloutAbort(t *testing.T) {
 	// The strategy emulates deployment all way down to 50/50 (steps 0 and 1)
 	for _, i := range []int{0, 1} {
 		step := vanguard.Steps[i]
-		t.Logf("setting release %q targetStep to %d", contenderName, i)
+		t.Logf("setting contender release %q targetStep to %d", contenderName, i)
 		f.targetStep(i, contenderName)
 
-		t.Logf("waiting for release %q to achieve waitingForCommand for targetStep %d", contenderName, i)
+		t.Logf("waiting for contender release %q to achieve waitingForCommand for targetStep %d", contenderName, i)
 		f.waitForReleaseStrategyState("command", contenderName, i)
 
 		expectedContenderCapacity := replicas.CalculateDesiredReplicaCount(uint(step.Capacity.Contender), float64(targetReplicas))
@@ -645,7 +645,7 @@ func TestRolloutAbort(t *testing.T) {
 	}
 
 	// The test emulates an interruption in the middle of the rollout, which
-	// means the incumbant becomes a new contender and it will stay in 50%
+	// means the incumbent becomes a new contender and it will stay in 50%
 	// capacity state (step 1 according to the vanguard definition) for a bit
 	// until shipper detects the need for capacity and spins up the missing
 	// pods
@@ -748,7 +748,7 @@ func (f *fixture) waitForRelease(appName string, historyIndex int) *shipper.Rele
 		f.t.Fatalf("error waiting for release to be scheduled: %q", err)
 	}
 
-	f.t.Logf("waiting for release took %s", time.Since(start))
+	f.t.Logf("waiting for release %q took %s", newRelease.Name, time.Since(start))
 	return newRelease
 }
 
@@ -811,12 +811,12 @@ func (f *fixture) waitForReleaseStrategyState(waitingFor string, releaseName str
 
 	if err != nil {
 		if err == wait.ErrWaitTimeout {
-			f.t.Fatalf("timed out waiting for release to be 'waitingForCommand': waited %s. final state: %s", globalTimeout, state)
+			f.t.Fatalf("timed out waiting for release to be waiting for %s: waited %s. final state: %s", waitingFor, globalTimeout, state)
 		}
-		f.t.Fatalf("error waiting for release to be 'waitingForCommand': %q", err)
+		f.t.Fatalf("error waiting for release to be waiting for %s: %q", waitingFor, err)
 	}
 
-	f.t.Logf("waiting for command took %s", time.Since(start))
+	f.t.Logf("waiting for %s took %s", waitingFor, time.Since(start))
 }
 
 func (f *fixture) waitForComplete(releaseName string) {
@@ -837,7 +837,7 @@ func (f *fixture) waitForComplete(releaseName string) {
 	if err != nil {
 		f.t.Fatalf("error waiting for release to complete: %q", err)
 	}
-	f.t.Logf("waiting for completion took %s", time.Since(start))
+	f.t.Logf("waiting for completion of %q took %s", releaseName, time.Since(start))
 }
 
 func poll(timeout time.Duration, waitCondition func() (bool, error)) error {

